uw-frame-components/portal/messages/controllers.js

```diff
@@ -0,0 +1,590 @@
+        vm.notificationsUrl = MESSAGES.notificationsPageURL;
+        vm.notificationsEnabled = MESSAGES.notificationsEnabled;
+        vm.status = 'View notifications';
+
+        // //////////////////
+        // Event listeners //
+        // //////////////////
+        /**
+         * When the parent controller has messages, initialize
+         * things dependent on messages
+         */
+        $scope.$watch('$parent.hasMessages', function(hasMessages) {
+          // If the parent scope has messages and notifications are enabled,
+          // complete initialization
+          if (hasMessages && vm.notificationsEnabled) {
+            configureNotificationsScope();
+          }
+        });
+
+        // ////////////////
+        // Local methods //
+        // ////////////////
+        /**
+         * Get notifications from parent scope, then pass them on
+         * for filtering by seen/unseen
+         */
+        var configureNotificationsScope = function() {
+          if ($scope.$parent.messages.notifications) {
+            allNotifications = $scope.$parent.messages.notifications;
+            // Get seen message IDs, then configure scope
+            $q.all(promiseSeenMessageIds)
+              .then(getSeenMessageIdsSuccess)
+              .catch(getSeenMessageIdsFailure);
+          }
+        };
+
+        /**
+         * Separate seen and unseen notifications, then set scope variables
+         * @param result
+         */
+        var getSeenMessageIdsSuccess = function(result) {
+          if (result.seenMessageIds && angular.isArray(result.seenMessageIds)
+            && result.seenMessageIds.length > 0) {
+            // Separate seen and unseen
+            separatedNotifications = $filter('filterSeenAndUnseen')(
+              allNotifications,
+              result.seenMessageIds
+            );
+            // Set scope notifications and dismissed notifications
+            vm.notifications = separatedNotifications.unseen;
+            vm.dismissedNotifications = separatedNotifications.seen;
+
+            // Set local dismissedNotificationIds (used for tracking
+            // dismissed messages in the K/V store
+            angular.forEach(vm.dismissedNotifications, function(value) {
+              dismissedNotificationIds.push(value.id);
+            });
+          } else {
+            // If there aren't any seen notification IDs, just set all
+            // notifications
+            vm.notifications = allNotifications;
+          }
+
+          // If user is anywhere other than the notifications page,
+          // check for priority notifications and set them in scope
+          if ($location.url().indexOf('notifications') === -1) {
+            configurePriorityNotificationsScope();
+          } else {
+            clearPriorityNotificationsFlags();
+          }
+
+          // Set aria-label in notifications bell
+          vm.status = 'You have '
+            + (vm.notifications.length === 0 ? 'no ' : '')
```

Comment:
Should this be telling the user how many notifications they have if it's nonzero?